### PR TITLE
Adding pictures with ActiveStorage

### PR DESCRIPTION
**Pull request by orhanna14**
This commit allows Users to add a picture to their profile and multiple pictures to their Recipes. 

Previously, I had introduced a Picture table, that was related to both User and Recipe and was going to build off of that. I instead decided to work with an ActiveStorage table with `active_storage_blobs` and `active_storage_attachments`., and build the relationship with the migration in this PR. (I apologize for the cleanliness of this commit history, I am working on this! :))

### CODE REVIEW

File: app/controllers/recipes_controller.rb [hunk 1]
```diff
@@ -17,16 +17,18 @@ def create
...
+      flash[:alert] = I18n.t('recipes.create.alert')
     end
   end
 
   private
 
   def recipe_params
-    params.require(:recipe).permit(:title, :ingredients, :instructions)
+    params.
+      require(:recipe).
```
[the remark this answers is not in the document]
Comment (orhanna14) [reply app/controllers/recipes_controller.rb]: I've got conflicting style issues from Hound and rubocop, I'm going to follow Rubocop on this one.

File: app/views/home/index.html.erb [hunk 2]
```diff
@@ -1 +1,9 @@
-<h1>Hello World!</h1>
+<h1>Hello <%= @user.name %></h1>
+<% if @user.picture.attached? %>
+  <p>
+  <strong>Picture:</strong>
+  <br>
```
Comment (mike-burns) [inline app/views/home/index.html.erb]: Maybe this should be a `dl`/`dt`/`dd` instead? Does that more closely match the semantics?

File: app/views/recipes/show.html.erb [hunk 3]
```diff
@@ -4,6 +4,17 @@
...
+        <strong>Pictures:</strong>
+        <br>
+        <%= @recipe.pictures.each do |picture| %>
+          <%= image_tag picture %>
+      <% end %>
+      </li>
+        </p>
   </ul>
-  <%= button_to t('.add_picture'), recipe_pictures_path(@recipe) %>
+<% end %>
```
Comment (mike-burns) [inline app/views/recipes/show.html.erb]: These tags seem mismatched. Is the generated HTML valid?

File: db/migrate/20191211174028_create_pictures.rb [hunk 4]
```diff
@@ -1,11 +0,0 @@
-# frozen_string_literal: true
-
-class CreatePictures < ActiveRecord::Migration[6.0]
-  def change
-    create_table :pictures do |t|
```
Comment (mike-burns) [inline db/migrate/20191211174028_create_pictures.rb]: Deleting this migration will cause the `DropPictures` migration to be invalid.

File: config/routes.rb [hunk 5]
```diff
@@ -3,9 +3,6 @@
 Rails.application.routes.draw do
   root 'home#index'
   resources :users, only: %i[new create] do
-    resources :recipes, only: %i[new create index show]
-  end
-  resources :recipes, only: %i[new create index show] do
-    resources :pictures, only: %i[new create]
+    resources :recipes, only: %i[new create edit index show]
```
Comment (mike-burns) [inline config/routes.rb]: Is there an `edit`?

File: app/views/home/index.html.erb [hunk 6]
```diff
@@ -1 +1,9 @@
...
+  <%= image_tag @user.picture %>
```
Comment (mike-burns) [inline app/views/home/index.html.erb]: Are there size constraints? Should this show a thumbnailed variant?

File: app/views/users/_form.html.erb [hunk 7]
```diff
@@ -12,3 +12,7 @@
   <%= form.label :password %>
   <%= form.password_field :password %>
 </div>
+
+<div class="picture-field">
```
Comment (mike-burns) [inline app/views/users/_form.html.erb]: Unclosed `div`.

File: spec/features/clearance/visitor_signs_up_spec.rb [hunk 8]
```diff
@@ -1,31 +1,43 @@
...
+  scenario 'with valid email, password, and picture' do
+    user = FactoryBot.build(:user)
+    user.
+      picture.
+      attach(io: File.open('public/apple-touch-icon.png'), filename: 'apple-touch-icon.png')
+
+    visit sign_up_path
+    click_button I18n.t('helpers.submit.user.create')
+
+    expect(user.picture.present?).to eq(true)
```
Comment (mike-burns) [inline spec/features/clearance/visitor_signs_up_spec.rb]: If you want to get fancy:

```suggestion
    expect(user.picture).to be_present
```

File: spec/system/signed_in_user_creates_a_recipe_spec.rb [hunk 9]
```diff
@@ -1,21 +1,39 @@
...
+    fill_in 'Instructions', with: recipe.instructions
+    click_button 'Create'
 
-    expect(page).to have_current_path("/")
+    expect(page).to have_current_path('/')
+  end
+
+  scenario 'and includes a picture' do
+    user = create(:user)
+    recipe = FactoryBot.build(:recipe)
```
Comment (mike-burns) [inline spec/system/signed_in_user_creates_a_recipe_spec.rb]: `FactoryBot.attributes_for` is maybe 0.00000000001 times faster than `FactoryBot.build`, in case you want that.

File: spec/system/signed_in_user_creates_a_recipe_spec.rb [hunk 10]
```diff
@@ -1,21 +1,39 @@
...
+
+    sign_in_with user.email, user.password
+    visit '/'
+    click_link I18n.t('layouts.application.new_recipe')
+    fill_in 'Title', with: recipe.title
```
Comment (mike-burns) [inline spec/system/signed_in_user_creates_a_recipe_spec.rb]: Should "Title", "Ingredients", etc. come from i18n?

File: spec/system/signed_in_user_creates_a_recipe_spec.rb [hunk 11]
```diff
@@ -1,21 +1,39 @@
...
+    fill_in 'Ingredients', with: recipe.ingredients
+    fill_in 'Instruction', with: recipe.instructions
+    recipe.
```
Comment (mike-burns) [inline spec/system/signed_in_user_creates_a_recipe_spec.rb]: Should this be [`#attach_file`](https://www.rubydoc.info/github/jnicklas/capybara/Capybara%2FNode%2FActions:attach_file)? Otherwise this is exercising ActiveStorage itself but not your app.

File: spec/features/clearance/visitor_signs_up_spec.rb [hunk 12]
```diff
@@ -1,31 +1,43 @@
...
+    sign_up_with 'testname', 'valid@example.com', 'password'
     visit sign_in_path
     expect_user_to_be_signed_in
   end
 
-  scenario "tries with invalid email" do
-    sign_up_with "testname", "invalid_email", "password"
+  scenario 'with valid email, password, and picture' do
+    user = FactoryBot.build(:user)
+    user.
```
Comment (mike-burns) [inline spec/features/clearance/visitor_signs_up_spec.rb]: Should probably use `#attach_file`.

File: app/views/recipes/show.html.erb (hunk 3 above)
Comment (stratigos) [reply app/views/recipes/show.html.erb]: I took a look at the whole file, and I am thinking that things would render more smoothly if the `</ul>` came _after_ the `<% end %>`. It appears as though if there are no pictures, then the closing tag wouldnt render either.

File: spec/features/clearance/visitor_signs_up_spec.rb [hunk 13]
```diff
@@ -1,31 +1,43 @@
...
-    click_link I18n.t("sessions.form.sign_up")
+    click_link I18n.t('sessions.form.sign_up')
 
     expect(current_path).to eq sign_up_path
   end
 
-  scenario "with valid email and password" do
-    sign_up_with "testname", "valid@example.com", "password"
+  scenario 'with valid email and password' do
+    sign_up_with 'testname', 'valid@example.com', 'password'
```
Comment (stratigos) [inline spec/features/clearance/visitor_signs_up_spec.rb]: Does the linter/style-tool youre using require single quotes? If so, I think thats perfectly fine in your own app, but I would like to point out the [thoughtbot style guide](https://github.com/thoughtbot/guides/tree/master/style/ruby) on this:

> Prefer double quotes for strings.

By no means is this a request for change :). But for client apps, we try to follow the above styles as much as we can.

File: config/routes.rb (hunk 5 above)
Comment (orhanna14) [reply config/routes.rb]: Oops, not yet. Thanks for pointing that out.

File: app/views/home/index.html.erb (hunk 2 above)
Comment (orhanna14) [reply app/views/home/index.html.erb]: Ah, maybe. I am unfamiliar with <dl/dt/dd>. I was focused only on functionality, but after looking into it, perhaps you're right!

File: db/migrate/20191211174028_create_pictures.rb (hunk 4 above)
Comment (orhanna14) [reply db/migrate/20191211174028_create_pictures.rb]: Thanks for pointing that out. I was unsure how to go about this. I created the `droptable` migration, dropped the model and the controller, and then ran the migration. 

I think removing this migration may have been part of my git cleanup. 
I'll keep that in mind for future migrations. Thanks!

File: spec/features/clearance/visitor_signs_up_spec.rb (hunk 13 above)
Comment (orhanna14) [reply spec/features/clearance/visitor_signs_up_spec.rb]: Thanks for pointing that out! I am actually getting conflicting messages from Hound and Rubocop. Hound says 'single-quote' and Rubocop says 'double-quote' I'll fix these to double quote.

File: app/views/home/index.html.erb (hunk 6 above)
Comment (orhanna14) [reply app/views/home/index.html.erb]: Definitely. Thanks!

File: spec/system/signed_in_user_creates_a_recipe_spec.rb (hunk 11 above)
Comment (orhanna14) [reply spec/system/signed_in_user_creates_a_recipe_spec.rb]: Yep, I agree! I initially tried `attach_file`, but was having trouble getting it to work when writing this. I will look into it more.

File: spec/system/signed_in_user_creates_a_recipe_spec.rb (hunk 10 above)
Comment (orhanna14) [reply spec/system/signed_in_user_creates_a_recipe_spec.rb]: I'm not sure. I was following  [our blog post here](https://thoughtbot.com/blog/testing-from-the-outsidein) and think I will leave this as is for now.

File: spec/system/signed_in_user_creates_a_recipe_spec.rb [hunk 14]
```diff
@@ -18,4 +18,23 @@
...
+    sign_in_with user.email, user.password
+    visit "/"
+    click_link I18n.t("layouts.application.new_recipe")
+    fill_in "Title", with: recipe.title
+    fill_in "Ingredients", with: recipe.ingredients
+    fill_in "Instruction", with: recipe.instructions
+
+    recipe.
+      pictures.
+      attach(io: File.open("public/apple-touch-icon.png"), filename: "apple-touch-icon.png")
```
[the remark this answers is not in the document]
Comment (orhanna14) [reply spec/system/signed_in_user_creates_a_recipe_spec.rb]: Ignoring this for now.

File: spec/features/clearance/visitor_signs_up_spec.rb [hunk 15]
```diff
@@ -18,6 +18,18 @@
     expect_user_to_be_signed_in
   end
 
+  scenario "with valid email, password, and picture" do
+    user = FactoryBot.build(:user)
+    user.
+      picture.
+      attach(io: File.open("public/apple-touch-icon.png"), filename: "apple-touch-icon.png")
```
[the remark this answers is not in the document]
Comment (orhanna14) [reply spec/features/clearance/visitor_signs_up_spec.rb]: Ignoring this for now.

File: app/views/home/index.html.erb [hunk 16]
```diff
@@ -1 +1,8 @@
-<h1>Hello World!</h1>
+<h1>Hello <%= @user.name %></h1>
+<% if @user.picture.attached? %>
+  <dl>
+  <dt>Picture:</dt>
+  <br>
+  <%= image_tag @user.picture.variant(resize: "500x500") %>
```
Comment (mike-burns) [inline app/views/home/index.html.erb]:
```html
<dd><%= image_tag @user.picture.variant(resize: "500x500") %></dd>
```

No need for the `<br>`.

**Comment by orhanna14** [comment]
Thanks for the input @mike-burns! This is a personal project (breakable toy) so I have not considered storage or what this would look like on a production environment.

File: app/views/recipes/show.html.erb [hunk 17]
```diff
@@ -4,6 +4,16 @@
   <ul>
     <li><%= @recipe.ingredients %></li>
     <li><%= @recipe.instructions %></li>
+    <% if @recipe.pictures.attached? %>
+      <li>
+        <p>
+        <strong>Pictures:</strong>
+        <br>
```
Comment (stratigos) [inline app/views/recipes/show.html.erb]: What would you think about controlling this break with css?

Alternatively, I think `<strong>Pictures:</strong>` and the `.each` loop could be placed inside their own `<div>` tags, which would likely display them on new lines. 

:memo: My inspiration comes from ideas like [this SO post](https://stackoverflow.com/questions/1726073/is-it-sometimes-bad-to-use-br). Id like to note that its from 2009, but I am assuming this is still valid in 2019.

File: db/migrate/20191203001550_add_clearance_to_users.rb [hunk 18]
```diff
@@ -10,7 +10,7 @@ def self.up
 
     add_index :users, :remember_token
 
-    users = select_all('SELECT id FROM users WHERE remember_token IS NULL')
+    users = select_all("SELECT id FROM users WHERE remember_token IS NULL")
```
Comment (stratigos) [inline db/migrate/20191203001550_add_clearance_to_users.rb]: I think the quote updates are fine to leave in this PR, especially since this is an app for your own experimentation. I would like to point out that for our client repositories, we would typically make stylistic changes like this in their own PR. This results in fewer things to think about when reviewing code, and allows for better management of the repo through smaller, focused changes.

File: spec/features/clearance/visitor_signs_up_spec.rb (hunk 12 above)
Comment (stratigos) [reply spec/features/clearance/visitor_signs_up_spec.rb]: :smile: Thanks for introducing me to `#attach_file`, @mike-burns. I read about it with [this Hashrocket post](https://til.hashrocket.com/posts/c790268652-attach-a-file-with-capybara) and the [rubydoc](https://www.rubydoc.info/github/jnicklas/capybara/Capybara%2FNode%2FActions%3Aattach_file). 

:thinking: Why would you suggest using that over `attach`? It has more simple arguments, but I am guessing the greater reason would be that it better reflects how a form would be used.

Adding: `attach_file("input-field-name", "public/apple-touch-icon.png")`  (and removing `user.picture.attach`) would allow capybara to test this more closely to the way a user would enter data on a page.

Any other reasons, or am I off base? TIA :)

Comment (stratigos) [reply spec/features/clearance/visitor_signs_up_spec.rb]: :file_folder: @orhanna14 What do you think of copying this image into a `spec/fixtures/` directory? I believe that using a resource in the `public/` dir makes the test more brittle as frontend resources could be swapped out or updated (renamed) at some point.

File: spec/system/signed_in_user_creates_a_recipe_spec.rb [hunk 19]
```diff
@@ -4,7 +4,7 @@
 require "support/features/clearance_helpers"
 
 RSpec.feature "A signed in user creates a recipe" do
-  scenario "with valid parameters" do
+  scenario "with valid parameters, but no picture" do
```
Comment (stratigos) [inline spec/system/signed_in_user_creates_a_recipe_spec.rb]: I was reading these specs in the file, and I am not sure if I understand the success condition. Does this imply a fail case if there is no picture? How does the location of "/" imply fail or success? Thanks in advance for filling me in :)

I see that the spec wasnt created for this PR, so perhaps if you determine an adjustment for this, it can come in a follow-up PR.

I was comparing with the new spec below, which has expectations that make succss/fail more obvious to me.

File: app/views/recipes/show.html.erb (hunk 17 above)
Comment (mike-burns) [reply app/views/recipes/show.html.erb]: I'd much rather see a `dl`/`dt`/`dd` than a `div`, personally. This is inline with [the upcoming accessibility guide](https://github.com/thoughtbot/guides/pull/579): Ensure semantic markup is used to describe content.

File: spec/features/clearance/visitor_signs_up_spec.rb (hunk 12 above)
Comment (mike-burns) [reply spec/features/clearance/visitor_signs_up_spec.rb]: Using `ActiveStore#attach` in the test is a test of ActiveStorage's `#attach` method; using Capybara's `#attach_file` method in the test is a test of the app. (I mentioned this more clearly [further down](https://github.com/orhanna14/homebrew-buddy/pull/21/files#r358995513), but `#attach_file` can be tricky to use.)

File: app/views/recipes/show.html.erb (hunk 17 above)
Comment (stratigos) [reply app/views/recipes/show.html.erb]: Thanks for bringing that to light - and I also agree :+1:

File: spec/system/signed_in_user_creates_a_recipe_spec.rb (hunk 19 above)
Comment (orhanna14) [reply spec/system/signed_in_user_creates_a_recipe_spec.rb]: Thanks for pointing that out. I updated it to read more clearly now, so that the reader knows the picture is optional.

File: db/migrate/20191203001550_add_clearance_to_users.rb (hunk 18 above)
Comment (orhanna14) [reply db/migrate/20191203001550_add_clearance_to_users.rb]: Great. Noted, thanks! I am definitely interested in improving my workflow/commit history to follow this. Be prepared for stylized PR's coming your way 👊

File: app/views/recipes/show.html.erb (hunk 17 above)
Comment (orhanna14) [reply app/views/recipes/show.html.erb]: Thanks for the insight. I was unfamiliar with `dl/dt/dd` previously, but this makes sense to have this file formatted as such.

File: spec/features/clearance/visitor_signs_up_spec.rb (hunk 12 above)
Comment (orhanna14) [reply spec/features/clearance/visitor_signs_up_spec.rb]: I was able to utilize `attach_file` and refactor the code a bit to create a `sign_up_with_picture` method in clearance helpers as well. 🎉  I also moved the image to the `spec/fixtures` directory. Thanks for the valuable input!

File: spec/system/signed_in_user_creates_a_recipe_spec.rb (hunk 11 above)
Comment (orhanna14) [reply spec/system/signed_in_user_creates_a_recipe_spec.rb]: I mentioned this in a comment above, but I managed to utilize `attach_file` here as well. 🎉 Thanks for pointing that out.

File: app/views/users/_form.html.erb (hunk 7 above)
Comment (orhanna14) [reply app/views/users/_form.html.erb]: That's embarrassing. Thanks 👍

File: spec/features/clearance/visitor_signs_up_spec.rb (hunk 12 above)
Comment (mike-burns) [reply spec/features/clearance/visitor_signs_up_spec.rb]: Oh heck yeah! I'm thrilled you got `attach_file` working!

**Comment by orhanna14** [comment]
I found an `active_storage_validation` [gem](https://github.com/igorkasyanchuk/active_storage_validations) and decided to implement simple validations of content type for now. I also currently want the picture to be optional, so I am allowing it to be nil for now as well. 

However, I would like to revisit this later, perhaps adding a default picture for user profiles and recipe when they choose to leave it blank.
